### PR TITLE
proposal for test folder consolidation

### PR DESCRIPTION
**Pull request by a-thaler**
Created on 2019-01-19 by Andreas Thaler (@Github a-thaler).

## Decision log

| Name | Description |
|-----------------------|------------------------------------------------------------------------------------|
| Title | Introduce a naming pattern for projects in the tests folder and apply the pattern for the existing projects. |
| Ultimate decision maker(s) | me |
| Due date | 2019-01-30 |
| Status | Proposed on 2019-01-19|
| Input provider(s) | All kyma contributors |
| Group(s) affected by the decision | All kyma contributors |
| Decision type | `Binary` |
| Earliest date to revisit the decision | 2019-02-01 |
| Affected decisions | `None`|

## Context
This decision proposes a change in the kyma-project/kyma/tests folder by proposing a consistent naming pattern for the projects inside that folder. New projects should follow the naming pattern and existing projects should be adjusted to it.

## Decision
A naming pattern needs to be agreed on and the decision to apply to it to all existing projects inside the tests folder.

## Consequences
It will cause a lot of renaming effort but will introduce a consistent naming and with that a better maintainability for the test projects.

### CODE REVIEW

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+- Have a readme in the _tests_ folder explaining the naming convention
+- Link the readme in the developer guide
+- Migrate _kubeless-integration_ to _e2e_ project containing e2e scenarios on a per package level resulting in one test suite to be executed.
+- All folders will stay in the root _tests_ folder but will be renamed according to the following table:
+
+| old folder | new folder | action required |
+|------------|------------|-----------------|
+| acceptance | dex | yes |
+| _ | service-catalog | yes |
+| _ | application-controller | yes |
```
[the remark this answers is not in the document]
Comment (a-thaler) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]: So you recommend to move it into the proposed e2e test suite, or should it be splitted somehow?

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# Test folder consolidation
+
+## Introduction
+The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains all tests executed against a running kyma cluster to assure integrity and functional correctness of the cluster when all modules are installed together.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+In kyma, that kind of tests are called _acceptance tests_.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
These are acceptance tests.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+All subfolders in the _tests_ directory define one test suite, usually focussing on one component.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
All subfolders in the `tests` directory belong to one test suite, usually focusing on one component.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+## Problem
+Every subfolder uses a different naming style, like _kubeless_ and _ui-api-layer-acceptance-tests_.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
Each subfolder uses a different naming style. For example, `kubeless` or `ui-api-layer-acceptance-tests`.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+With that:
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
As a result:
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+- Overall it is not easy to decide which convention to follow when creating a new subfolder.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- It's difficult to decide which convention to follow when creating a new subfolder.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+- Not always it is possible to derive the related component (there is no e2e test yet, they are all focussed on components).
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- It is not always possible to derive the related component (there is no e2e test yet, the tests are all focused on components).
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+- It is confusing for readers
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- It is confusing for readers.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+## Goal
+
+Propose a name pattern which places the valuable information, the differentiator of the folder, into the focus.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
Propose a name pattern using specific information to distinguish between the folders.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+## Consideration
+
+**kind of test** All tests are of the same style currently, named as _acceptance tests_. As long as there is no different kind of tests available, the information of the kind of test is not valuable as part of the subfolder name
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**kind of test** Currently, all tests are called **acceptance tests**. As long as there is no different kind of tests available, the test kind is not valuable information as part of the subfolder name.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 12]
```diff
@@ -0,0 +1,67 @@
...
+
+**test prefix/suffix** all subfolders contain test suites. That's why they are located under the _tests_ folder. Having a prefix or suffix in the folder name is redundant and does not bring more value to the reader.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**test prefix/suffix** All subfolders contain test suites. That's why they reside in the `test` folder. A prefix or suffix in the folder name is redundant and does not bring more value to the reader.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+**one component per folder** as long as a test suite is focussing on a specific component under test, the test suite should be named accordingly. With that the component and also the maintainers can be identified more quickly.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**one component per folder** As long as a test suite focuses on a specific component, the test suite should be named accordingly. This way the component and its maintainers can be identified quicker.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+There is one special folder called _acceptance_ currently containing test suites for 3 different components, but sharing some few common scripts. As the tendency is to have a dedicated folder and docker image per component, that folder should be converted into three dedicated component folders with own docker images. Common scripts can be still shared on the root level or a dedicated _common_ folder.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
The `acceptance` folder currently contains test suites for 3 different components. These suites share some few common scripts. The best practice is to have a dedicated folder per component, including a docker image. That's why the `acceptance` folder should be converted into three dedicated component folders with their own docker images. Common scripts can be still shared on the root level or in a `common` folder.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+
+**docker images** Every subfolder is producing a docker image using the folder name as image name. If no prefix/suffix is used, the name will collide with the image of the actual component. To have a differentiation here, a subfolder for the image could be introduced
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**docker images** Each subfolder produces a docker image using the folder name for the image name. If no prefix/suffix is used, the folder and image names can be conflicting. Introducing a subfolder for the image could solve this problem. 
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+
+## Proposal
+
+The subfolder specific to a component should be named like the component, that's it.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
Use the component's name for the component-specific subfolder.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+
+No prefix, no suffix, no _acceptance_, no multi-components
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
No prefix, no suffix, no `acceptance` no multi-components.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+
+The resulting docker image of a subfolder should be located in the subfolder _tests_
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
The docker image for a subfolder should reside in the `tests` subfolder.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+**Example**: A component _event-bus_ has its acceptance tests in folder _tests/event-bus_ and produces a docker image _XX/tests/event-bus:0.5.1_
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces a docker image located in `XX/tests/event-bus:0.5.1`
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+
+Real e2e scenarios (like kubeles--integration) should be bundled into one subfolder _e2e_. Here we should have one test project which will execute all e2e tests organized by scenarios in different packages.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
Bundle the real e2e scenarios (like Kubeless--integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios to different packages.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 21]
```diff
@@ -0,0 +1,67 @@
...
+
+## Concrete Actions
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
## Actions
```
Or Action Points

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+
+- Have a readme in the _tests_ folder explaining the naming convention
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- Include a `README.md` file in the `tests` folder to explain the naming convention.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 23]
```diff
@@ -0,0 +1,67 @@
...
+- Link the readme in the developer guide
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- Link the `README.md` file in the developer guide.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 24]
```diff
@@ -0,0 +1,67 @@
...
+- Migrate _kubeless-integration_ to _e2e_ project containing e2e scenarios on a per package level resulting in one test suite to be executed.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- Migrate the `kubeless-integration_ to _e2e` project containing e2e scenarios on the package level to have one test suite.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 25]
```diff
@@ -0,0 +1,67 @@
...
+- All folders will stay in the root _tests_ folder but will be renamed according to the following table:
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- Keep all folders in the root `tests` folder rename them according to the information in this table:
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 4 above)
Comment (a-thaler) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]: One folder means one test suite, so they are really defining it
```suggestion
All subfolders in the `tests` directory define one test suite, usually focusing on one component.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 19 above)
Comment (a-thaler) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces a docker image named like `XX/tests/event-bus:0.5.1`
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 20 above)
Comment (a-thaler) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
+ Bundle the real e2e scenarios (like kubeless-integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios to different packages.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 26]
```diff
@@ -0,0 +1,67 @@
+# Test folder consolidation
+
+## Introduction
+The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
The [acceptance tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.
```
I'm wondering if we maybe can merge the line saying `these are acceptance tests` - in the link name?

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+These are acceptance tests.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion

```

Remove the line and merge it with the paragraph above.

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+All subfolders in the `tests` directory define one test suite, usually focusing on one component.
+
+## Problem
+Each subfolder uses a different naming style. For example, `kubeless` or `ui-api-layer-acceptance-tests`.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
Each subfolder uses a different naming pattern. For example, `kubeless` or `ui-api-layer-acceptance-tests`.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 29]
```diff
@@ -0,0 +1,67 @@
...
+
+As a result:
+- It's difficult to decide which convention to follow when creating a new subfolder.
+- It is not always possible to derive the related component (there is no e2e test yet, the tests are all focused on components).
+- It is confusing for readers.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- It's confusing for readers.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 30]
```diff
@@ -0,0 +1,67 @@
...
+The [tests](https://github.com/kyma-project/kyma/tree/master/tests) folder of the kyma-project contains tests executed against a running Kyma cluster to assure integrity and functional correctness of the cluster with all modules installed.
+These are acceptance tests.
+All subfolders in the `tests` directory define one test suite, usually focusing on one component.
+
+## Problem
+Each subfolder uses a different naming style. For example, `kubeless` or `ui-api-layer-acceptance-tests`.
+
+As a result:
+- It's difficult to decide which convention to follow when creating a new subfolder.
+- It is not always possible to derive the related component (there is no e2e test yet, the tests are all focused on components).
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- It's not always possible to derive the related component (there is no e2e test yet, the tests are focused on components).
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 31]
```diff
@@ -0,0 +1,67 @@
...
+
+**docker images** Each subfolder produces a docker image using the folder name for the image name. If no prefix/suffix is used, the folder and image names can be conflicting. Introducing a subfolder for the image could solve this problem. 
+
+## Proposal
+
+Use the component's name for the component-specific subfolder.
+
+No prefix, no suffix, no `acceptance` no multi-components.
+
+The docker image for a subfolder should reside in the `tests` subfolder.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
The docker image for a subfolder resides in the `tests` subfolder.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 32]
```diff
@@ -0,0 +1,67 @@
...
+
+**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces a docker image named like `XX/tests/event-bus:0.5.1`
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
**Example**: The Event-Bus component has its acceptance tests in the `tests/event-bus` folder and produces the `XX/tests/event-bus:0.5.1` docker image.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 33]
```diff
@@ -0,0 +1,67 @@
...
+
++ Bundle the real e2e scenarios (like kubeless-integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios to different packages.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
+ Bundle the real e2e scenarios (like kubeless-integration) into one `e2e` subfolder. Here we should have one test project which executes all e2e tests divided by scenarios into different packages.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 34]
```diff
@@ -0,0 +1,67 @@
...
+
+## Actions
+
+- Include a `README.md` file in the `tests` folder to explain the naming convention.
+- Link the `README.md` file in the developer guide.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- Link the `README.md` file in the developer's guide.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md [hunk 35]
```diff
@@ -0,0 +1,67 @@
...
+- Migrate the `kubeless-integration_ to _e2e` project containing e2e scenarios on the package level to have one test suite.
```
Comment (bszwarc) [inline sig-and-wg/sig-core/proposals/test-folder-consolidation.md]:
```suggestion
- Migrate the `kubeless-integration` to `e2e` project containing e2e scenarios on the package level to have one test suite.
```

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 26 above)
Comment (a-thaler) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]: The folder is called "tests", not acceptance tests

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 27 above)
Comment (a-thaler) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]: No, I don't like the idea. We are talking about the tests contained in that folder. I want to express that these tests are covering one test suite each and they act in the role of acceptance tests. That consideration should be the outcome of the paragraph for the reader.

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 26 above)
Comment (bszwarc) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]: Ok - I get it. Let's leave it as it is.

File: sig-and-wg/sig-core/proposals/test-folder-consolidation.md (hunk 27 above)
Comment (bszwarc) [reply sig-and-wg/sig-core/proposals/test-folder-consolidation.md]: Ok, sure - let's leave it then.

**Pull request merged**